3106 gov ct2

martinscripts opened the pull request:
hey @joyelfeghali @adriguir,
I did not find the time to get the NRT to run, but I will work on it soon. 
Please already review the added models if you have time.

BR,
.Martin

**Checklist before requesting a review**

*use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes*

- [x] unit tests and non-regression tests were added (new model, new feature and bug fix)
- [ ] main documentation was updated (update of input/output file, 3rd party, model, repository organization, solver)
- [ ] example documentations were updated (new example in examples folder)
- [ ] the corresponding milestone was added in the ticket and in this PR

joyelfeghali commented on nrt/data/SMIB/Standard/TestCaseGovCT2/curves.crv:
After getting the results, you should add them to a reference folder for curves. (you copy the file TestCaseGovCT2/outputs/curves/curves.crv)

joyelfeghali commented on dynawo/sources/Models/Modelica/PreassembledModels/CMakeLists.txt:
```diff
@@ -61,6 +61,7 @@ set(MODELICA_PREASSEMBLED_MODELS
   GeneratorPVRemoteDiagramPQSignalN.xml
   GeneratorPVRemoteDiagramPQSignalNSFR.xml
   GeneratorSynchronousFourWindings.xml
+  GeneratorSynchronousFourWindingsGovCT2SexsPss2A.xml
```
Also could you add this file in PreassembledModels/NrtModels.txt ?

joyelfeghali commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  parameter Types.PerUnit KPGovPu = 4 "Governor proportional gain in pu" annotation( Dialog(tab = "Main control path"));
+  parameter Types.PerUnit KPLoadPu = 1 "Load limiter proportional gain for PI controller in pu" annotation( Dialog(tab = "Load limit controller"));
+  parameter Types.PerUnit KTurbPu = 1.9168 "Turbine gain in pu" annotation( Dialog(tab = "Turbine/engine"));
+  parameter Types.PerUnit omega0Pu = SystemBase.omega0Pu "Initial rotor speed";
```
Types.AngularVelocityPu

martinscripts replied:
done

joyelfeghali commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  parameter Types.Time ValveMaxPu = 1 "Maximum valve position limit in pu" annotation( Dialog(tab = "Turbine/engine"));
+  parameter Types.Time ValveMinPu = 0.175 "Minimum valve position limit in pu" annotation( Dialog(tab = "Turbine/engine"));
+  parameter Types.Time WFnlPu = 0.187 "No load fuel flow in pu" annotation( Dialog(tab = "Turbine/engine"));
+  parameter Boolean WFSpdBool = false "Switch for fuel source characteristic" annotation( Dialog(tab = "Turbine/engine")); 
```
You gave a default value for all these parameters, it is good when using the model in OpenModelica, but in dynawo if we try to modify a default value in the .par, it might not be taken into account. This is something we are working on.
But I would suggest removing these default values and that you add in the description of every parameter the "typical value" similar to what we do in Dynawo.Electrical.Controls.WECC.Parameters.ParamsElectricalControl

martinscripts replied:
done

joyelfeghali commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  parameter Boolean WFSpdBool = false "Switch for fuel source characteristic" annotation( Dialog(tab = "Turbine/engine")); 
+//
+  // initialization helpers
+  final parameter Real initCfe = WFnlPu + (if KTurbPu > 0 then initPMechNoLoss/KTurbPu else 0);
```
Types.PerUnit?

martinscripts replied:
done

joyelfeghali commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  final parameter Types.PerUnit initFsrt = (PLdRefPu/KTurbPu + WFnlPu - initTex)*KPLoadPu + initIntegratorKILoad;
+  final parameter Real initIntegratorKILoad = 1;
+  final parameter Types.PerUnit initPMechNoLoss = if DmPu>0.0 then Pm0Pu + omega0Pu*DmPu else Pm0Pu;
+  final parameter Real initTex = initCfe * (if DmPu<0 then omega0Pu^(-DmPu) else 1);
```
Types.PerUnit

joyelfeghali commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+    Line(points = {{-332, 24}, {-280, 24}, {-280, -12}, {-256, -12}}, color = {0, 0, 127}));
+  annotation(
+    preferredView = "diagram",
+    uses(Modelica(version = "3.2.3")),
```
we could remove the modelica version in here

joyelfeghali commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  Modelica.Blocks.Sources.RealExpression omegaPu2(y = omegaPu) annotation( Placement(visible = true, transformation(origin = {157, 89}, extent = {{13, -9}, {-13, 9}}, rotation = 0)));
+  Modelica.Blocks.Sources.RealExpression omegaPu3(y = omegaPu) annotation( Placement(visible = true, transformation(origin = {341, -157}, extent = {{13, -9}, {-13, 9}}, rotation = 0)));
+  Modelica.Blocks.Sources.RealExpression omegaPu4(y = omegaPu) annotation( Placement(visible = true, transformation(origin = {149, 227}, extent = {{-13, -9}, {13, 9}}, rotation = 0)));
+  Modelica.Blocks.Interfaces.RealOutput PmPu(start = Pm0Pu) annotation( Placement(visible = true, transformation(origin = {330, 132}, extent = {{-10, -10}, {10, 10}}, rotation = 0), iconTransformation(origin = {350, 2}, extent = {{-30, -30}, {30, 30}}, rotation = 0)));
```
description + base PNomTurb

joyelfeghali commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  // Inputs, Outputs
+  Modelica.Blocks.Interfaces.RealInput omegaPu(start = SystemBase.omega0Pu) annotation( Placement(visible = true, transformation(origin = {-329, -19}, extent = {{-13, -13}, {13, 13}}, rotation = 0), iconTransformation(origin = {-344, 100}, extent = {{-24, -24}, {24, 24}}, rotation = 0)));
+  Modelica.Blocks.Interfaces.RealInput omegaRefPu(start = SystemBase.omega0Pu) annotation(Placement(visible = true, transformation(origin = {-333, 23}, extent = {{-13, -13}, {13, 13}}, rotation = 0), iconTransformation(origin = {-343, 181}, extent = {{-23, -23}, {23, 23}}, rotation = 0)));
+  Modelica.Blocks.Interfaces.RealInput PePu(start = Pm0Pu) annotation( Placement(visible = true, transformation(origin = {-335, -183}, extent = {{-13, -13}, {13, 13}}, rotation = 0), iconTransformation(origin = {-343, -161}, extent = {{-23, -23}, {23, 23}}, rotation = 0)));
```
The init model would be : 
within Dynawo.Electrical.Controls.Machines.Governors;
/*...
*/

model Governor2_INIT "Second initialization model for governors"
  extends AdditionalIcons.Init;

  Types.ActivePowerPu PGen0Pu "Initial active power in pu (base SnRef) (generator convention)";
  Types.ActivePowerPu Pm0Pu "Initial mechanical power in pu (base PNomTurb)";

  annotation(preferredView = "text");
end Governor2_INIT;

martinscripts replied:
done

joyelfeghali commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+    experiment(StartTime = 0, StopTime = 10.5, Tolerance = 0.001, Interval = 0.002),
+    __OpenModelica_simulationFlags(lv = "LOG_STATS", s = "rungekutta"),
+    __OpenModelica_commandLineOptions = "--matchingAlgorithm=PFPlusExt --indexReductionMethod=dynamicStateSelection -d=initialization,NLSanalyticJacobian");
+end GovCT2;
```
Could you organise the model as in drawing a rectangle to specify the three parts : 
- main control path  (resulting in fsrn)
- Acceleration command (resulting in fsra)
- Temperature command (resulting in fsrt)

You could also increase the white Diagram size to have more space

joyelfeghali commented on dynawo/sources/Models/Modelica/PreassembledModels/GeneratorSynchronousFourWindingsGovCT2SexsPss2A.xml:
```diff
@@ -0,0 +1,34 @@
+  <dyn:modelicaModel id="GeneratorSynchronousFourWindingsGovCT2SexsPss2A">
+    <dyn:unitDynamicModel id="generator" name="Dynawo.Electrical.Machines.OmegaRef.GeneratorSynchronous" initName="Dynawo.Electrical.Machines.OmegaRef.GeneratorSynchronousExt4W_INIT"/>
+    <dyn:unitDynamicModel id="voltageRegulator" name="Dynawo.Electrical.Controls.Machines.VoltageRegulators.Standard.SEXS" initName="Dynawo.Electrical.Controls.Machines.VoltageRegulators.Standard.SEXS_INIT"/>
+    <dyn:unitDynamicModel id="governor" name="Dynawo.Electrical.Controls.Machines.Governors.Standard.Generic.GovCT2" initName="Dynawo.Electrical.Controls.Machines.Governors.Governor_INIT"/>
```
If the modifications are taken into account, the init model should be changed here

martinscripts replied:
done

martinscripts commented:
@joyelfeghali, 
the NRT runs now. I had to change to simple solver. 
I also added the reference/curves.csv

martinscripts commented:
@joyelfeghali 
I added a commit that retroactively adds signoffs according to [this}(https://wiki.lfenergy.org/display/HOME/Contribution+and+Compliance+Guidelines).

Please let me know if this is sufficient, since the DCO check still isn't positive

martinscripts commented:
@joyelfeghali 
I added the missing signoffs to past_commits.txt as it was described in the link that you supplied, ([lfenergy guidelines](https://wiki.lfenergy.org/display/HOME/Contribution+and+Compliance+Guidelines)). 
Shouldn't this work?

joyelfeghali commented:
> @joyelfeghali I squashed and rebased just now.

great thanks, it seems that the signed-off works now.

Just one last thing about the commit : it should start with "#3106" Usually it is done automatically when you work on dynawo repository (you can see other PR).
Could add this number at the beginning of he commit name?

martinscripts commented:
> > @joyelfeghali I squashed and rebased just now.
> 
> great thanks, it seems that the signed-off works now.
> 
> Just one last thing about the commit : it should start with "#3106" Usually it is done automatically when you work on dynawo repository (you can see other PR). Could add this number at the beginning of he commit name?

Done @joyelfeghali

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  Modelica.Blocks.Interfaces.RealInput omegaRefPu(start = SystemBase.omega0Pu) annotation(Placement(visible = true, transformation(origin = {-333, 23}, extent = {{-13, -13}, {13, 13}}, rotation = 0), iconTransformation(origin = {-343, 181}, extent = {{-23, -23}, {23, 23}}, rotation = 0)));
+  Modelica.Blocks.Interfaces.RealInput PePu(start = Pm0Pu) annotation( Placement(visible = true, transformation(origin = {-335, -183}, extent = {{-13, -13}, {13, 13}}, rotation = 0), iconTransformation(origin = {-343, -161}, extent = {{-23, -23}, {23, 23}}, rotation = 0)));
+  Modelica.Blocks.Interfaces.RealInput PMwSetPu(start = PRef0Pu) annotation( Placement(visible = true, transformation(origin = {-333, -139}, extent = {{-13, -13}, {13, 13}}, rotation = 0), iconTransformation(origin = {-344, -80}, extent = {{-24, -24}, {24, 24}}, rotation = 0)));
+  Modelica.Blocks.Interfaces.RealInput PRefPu(start = PRef0Pu) annotation( Placement(visible = true, transformation(origin = {-333, -87}, extent = {{-13, -13}, {13, 13}}, rotation = 0), iconTransformation(origin = {-346, 10}, extent = {{-24, -24}, {24, 24}}, rotation = 0)));
```
neither, it depends on the control mode. if RSelect=2, for example, it is the valve positon. And it is multiplied by R as well

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  Modelica.Blocks.Math.Gain gainR(k = RPu) annotation( Placement(visible = true, transformation(origin = {-172, -124}, extent = {{10, -10}, {-10, 10}}, rotation = -90)));
+  Modelica.Blocks.Continuous.Integrator integratorKIGov(initType = Modelica.Blocks.Types.Init.InitialState, k = KIGovPu, y_start = initValve) annotation( Placement(visible = true, transformation(origin = {-50, -92}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
+  Modelica.Blocks.Continuous.Integrator integratorKILoad(initType = Modelica.Blocks.Types.Init.InitialState, k = KILoadPu, y_start = initIntegratorKILoad) annotation( Placement(visible = true, transformation(origin = {-104, 34}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
+  Dynawo.NonElectrical.Blocks.NonLinear.LimitedIntegrator integratorKIMw(K = KIMwPu, YMax = 1.1*RPu, YMin = -1.1*RPu) annotation( Placement(visible = true, transformation(origin = {-238, -120}, extent = {{10, -10}, {-10, 10}}, rotation = -90)));
```
![image](https://github.com/user-attachments/assets/eace5eab-ceb3-4b13-bd8d-0ec05a61b0f0)

Pref is initialized to whatever comes out of the "R" gain block, so I would argue that Y0=0 is fine for the KIMW integrator, at least if omegaRef=omega at initial state, which I would assume. No?

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  final parameter Types.PerUnit initPMechNoLoss = if DmPu>0.0 then Pm0Pu + omega0Pu*DmPu else Pm0Pu;
+  final parameter Real initTex = initCfe * (if DmPu<0 then omega0Pu^(-DmPu) else 1);
+  final parameter Real initValve = if WFSpdBool then initCfe/omega0Pu else initCfe;
+  final parameter Types.PerUnit PRef0Pu = RPu*(if RSelectInt == 0 then 0 else if RSelectInt == 1 then Pm0Pu else initValve) "Initial value of reference power";
```
true! Done.

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  Modelica.Blocks.Logical.Switch switchWFSpd annotation( Placement(visible = true, transformation(origin = {272, -126}, extent = {{10, -10}, {-10, 10}}, rotation = -90)));
+  Modelica.Blocks.Tables.CombiTable1Ds tablePLimFromf(extrapolation = Modelica.Blocks.Types.Extrapolation.LastTwoPoints, smoothness = Modelica.Blocks.Types.Smoothness.LinearSegments, table = PLimFromfPoints, tableOnFile = false, verboseRead = false) annotation( Placement(visible = true, transformation(origin = {138, 38}, extent = {{-10, -10}, {10, 10}}, rotation = -90)));
+  Modelica.Blocks.Continuous.TransferFunction transferFuncKDGovTDGov(a = {tDGovSeconds, 1}, b = {KDGovPu, 0}, initType = Modelica.Blocks.Types.Init.InitialState) annotation( Placement(visible = true, transformation(origin = {-49, -133}, extent = {{-13, -13}, {13, 13}}, rotation = 0)));
+  Modelica.Blocks.Continuous.TransferFunction transferFunctA(a = {tASeconds, 1}, b = {1, 0}) annotation( Placement(visible = true, transformation(origin = {-142, -18}, extent = {{-12, -12}, {12, 12}}, rotation = 0)));
```
done

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  final parameter Types.PerUnit initFsrt = (PLdRefPu/KTurbPu + WFnlPu - initTex)*KPLoadPu + initIntegratorKILoad;
+  final parameter Real initIntegratorKILoad = 1;
+  final parameter Types.PerUnit initPMechNoLoss = if DmPu>0.0 then Pm0Pu + omega0Pu*DmPu else Pm0Pu;
+  final parameter Real initTex = initCfe * (if DmPu<0 then omega0Pu^(-DmPu) else 1);
```
I think you are correct, thank you. done

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  parameter Types.PerUnit DeltaOmegaDbPu = 0 "Speed governor deadband in PU speed." annotation( Dialog(tab = "Main control path"));
+  parameter Types.PerUnit DeltaOmegaMaxPu = 1 "Maximum value for speed error signal in pu" annotation( Dialog(tab = "Main control path"));
+  parameter Types.PerUnit DeltaOmegaMinPu = -1 "Minimum value for speed error signal in pu" annotation( Dialog(tab = "Main control path"));
+  parameter Types.Time DeltaTSeconds = 1 "Correction factor in s (to adapt the unit of the acceleration limiter gain from pu/s to pu)" annotation( Dialog(tab = "Acceleration limiter"));
```
done

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  parameter Types.Time tSBSeconds = 50 "Temperature detection lag time constant in s" annotation( Dialog(tab = "Load limit controller"));
+  parameter Types.Time ValveMaxPu = 1 "Maximum valve position limit in pu" annotation( Dialog(tab = "Turbine/engine"));
+  parameter Types.Time ValveMinPu = 0.175 "Minimum valve position limit in pu" annotation( Dialog(tab = "Turbine/engine"));
+  parameter Types.Time WFnlPu = 0.187 "No load fuel flow in pu" annotation( Dialog(tab = "Turbine/engine"));
```
done

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  Dynawo.NonElectrical.Blocks.Continuous.RateLimFirstOrderFreeze firstOrdertActuatorRatelim(T = tActuatorSeconds, UseRateLim = true, Y0 = initValve, y(fixed = true)) annotation( Placement(visible = true, transformation(origin = {214, -88}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
+  Modelica.Blocks.Continuous.FirstOrder firstOrdertFLoad(T = tFLoadSeconds, initType = Modelica.Blocks.Types.Init.InitialOutput, y_start = initTex) annotation( Placement(visible = true, transformation(origin = {34, 114}, extent = {{10, -10}, {-10, 10}}, rotation = 0)));
+  Modelica.Blocks.Continuous.FirstOrder firstOrdertPElec(T = tPElecSeconds, initType = Modelica.Blocks.Types.Init.InitialOutput, y_start = Pm0Pu) annotation( Placement(visible = true, transformation(origin = {-250, -182}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
+  Modelica.Blocks.Math.Gain gainfNom(k(unit = "") = fNomHz) annotation( Placement(visible = true, transformation(origin = {138, 66}, extent = {{10, -10}, {-10, 10}}, rotation = 90)));
```
true, done

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  parameter Types.Frequency fLim6Hz = 4 "Frequency threshold 6 in Hz" annotation( Dialog(tab = "Frequency dependent valve limit"));
+  parameter Types.Frequency fLim7Hz = 3 "Frequency threshold 7 in Hz" annotation( Dialog(tab = "Frequency dependent valve limit"));
+  parameter Types.Frequency fLim8Hz = 2 "Frequency threshold 8 in Hz" annotation( Dialog(tab = "Frequency dependent valve limit"));
+  parameter Types.Frequency fLim9Hz = 1 "Frequency threshold 9 in Hz" annotation( Dialog(tab = "Frequency dependent valve limit"));
```
exactly, they shouldnt be reached. It is a flat line characteristic as default

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  parameter Types.Frequency fLim7Hz = 3 "Frequency threshold 7 in Hz" annotation( Dialog(tab = "Frequency dependent valve limit"));
+  parameter Types.Frequency fLim8Hz = 2 "Frequency threshold 8 in Hz" annotation( Dialog(tab = "Frequency dependent valve limit"));
+  parameter Types.Frequency fLim9Hz = 1 "Frequency threshold 9 in Hz" annotation( Dialog(tab = "Frequency dependent valve limit"));
+  parameter Types.Frequency fNomHz = 50 "Nominal Frequency in Hz";
```
done

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  parameter Types.Time WFnlPu = 0.187 "No load fuel flow in pu" annotation( Dialog(tab = "Turbine/engine"));
+  parameter Boolean WFSpdBool = false "Switch for fuel source characteristic" annotation( Dialog(tab = "Turbine/engine")); 
+//
+  // initialization helpers
```
done

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+    Line(points = {{156, 158}, {174, 158}, {174, 144}, {202, 144}}, color = {0, 0, 127}));
+  connect(omegaRefPu, add.u1) annotation(
+    Line(points = {{-332, 24}, {-280, 24}, {-280, -12}, {-256, -12}}, color = {0, 0, 127}));
+  annotation(
```
done

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/package.mo:
```diff
@@ -1,20 +1,25 @@
 
-  annotation(preferredView = "info",
-    Documentation(info = "<html><head></head><body>This package contains different governors models and is organized into different subpackages:<div><ul><li><a href=\"modelica://Dynawo.Electrical.Controls.Machines.Governors.Standard.Steam\">\"Steam\"</a> for steam standard governor models </li></ul></div></body></html>"));
+
```
done

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/package.mo:
```diff
@@ -0,0 +1,21 @@
+within Dynawo.Electrical.Controls.Machines.Governors.Standard;
+
+package Generic "Generic standard governors"
+  /*
```
done

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/package.mo:
```diff
@@ -1,20 +1,25 @@
-
 package Standard "Standard governors"
+  /*
+  * Copyright (c) 2021, RTE (http://www.rte-france.com)
```
very strange that this even moved. I have the suspicion that OMEdit auto-formats things too much sometimes.

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/PowerExternalBase.mo:
```diff
@@ -0,0 +1,24 @@
+  
+equation
+  y = u1^u2;
+  annotation(
```
done

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/PowerExternalBase.mo:
No, because in Math.Power, the parameter is the base (standard = e) and the input signal is the exponent. 
--> e^u1

but we want the base to be variable.

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,357 @@
+  parameter Types.PerUnit KILoad = 1 "Load limiter integral gain for PI controller, typical: 1" annotation( Dialog(tab = "Load limit controller"));
+  parameter Types.PerUnit KIMw = 0 "Power controller gain, typical: 0 (deactivated)" annotation( Dialog(tab = "Supervisory load controller"));
+  parameter Types.PerUnit KPGov = 4 "Governor proportional gain, typical: 4" annotation( Dialog(tab = "Main control path"));
+  parameter Types.PerUnit KPLoad = 1 "Load limiter proportional gain for PI controller, typical: 1" annotation( Dialog(tab = "Load limit controller"));
```
`In the model schema there is no KPLoad but KILoad instead is this normal?` 
Thanks, I fixed this. 

`They are always equal?`
no, but they are both given 1 as standard values in the standard

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,272 @@
+  parameter Types.PerUnit RDownPu = -99 "Maximum rate of load limit decrease in pu/s (base PNomTurb*KTurb), typical value = -99" annotation( Dialog(tab = "Load limit controller"));
+  parameter Types.PerUnit ROpenPu = 99 "Maximum valve opening rate in pu/s (base PNomTurb*KTurb), typical value = 99" annotation( Dialog(tab = "Turbine/engine"));
+  parameter Types.PerUnit RDroop = 0.05 "Permanent droop, typical value = 0.05" annotation( Dialog(tab = "Main control path"));
+  parameter Integer RSelectInt "Feedback signal for droop" annotation( Dialog(tab = "Main control path"));
```
done

martinscripts commented on dynawo/sources/Models/Modelica/PreassembledModels/GeneratorSynchronousFourWindingsGovCT2SexsPss2A.xml:
```diff
@@ -0,0 +1,35 @@
+  <!-- Machine with four windings and standard regulations - GovCT2, SEXS and PSS2A-->
+  <dyn:modelicaModel id="GeneratorSynchronousFourWindingsGovCT2SexsPss2A">
+    <dyn:unitDynamicModel id="generator" name="Dynawo.Electrical.Machines.OmegaRef.GeneratorSynchronous" initName="Dynawo.Electrical.Machines.OmegaRef.GeneratorSynchronousExt4W_INIT"/>
+    <dyn:unitDynamicModel id="voltageRegulator" name="Dynawo.Electrical.Controls.Machines.VoltageRegulators.Standard.SEXS" initName="Dynawo.Electrical.Controls.Machines.VoltageRegulators.Standard.SEXS_INIT"/>
```
done

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,272 @@
+within Dynawo.Electrical.Controls.Machines.Governors.Standard.Generic;
+
+/*
+* Copyright (c) 2024, RTE (http://www.rte-france.com) and UPC/Citcea (https://www.citcea.upc.edu/)
```
done

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,272 @@
+  Modelica.Blocks.Tables.CombiTable1Ds tablePLimFromf(extrapolation = Modelica.Blocks.Types.Extrapolation.LastTwoPoints, smoothness = Modelica.Blocks.Types.Smoothness.LinearSegments, table = PLimFromfPoints, tableOnFile = false, verboseRead = false) annotation( Placement(visible = true, transformation(origin = {138, 38}, extent = {{-10, -10}, {10, 10}}, rotation = -90)));
+  Modelica.Blocks.Continuous.TransferFunction transferFunctCtB(a = {tB, 1}, b = {tC, 1}, initType = Modelica.Blocks.Types.Init.InitialState, x_start = {initPMechNoLossPu}, y_start = initPMechNoLossPu) annotation( Placement(visible = true, transformation(origin = {254, 106}, extent = {{10, -10}, {-10, 10}}, rotation = -90)));
+  Modelica.Blocks.Continuous.TransferFunction transferFunctSAtSB(a = {tSB, 1}, b = {tSA, 1}, initType = Modelica.Blocks.Types.Init.InitialOutput, y_start = initTexPu) annotation( Placement(visible = true, transformation(origin = {80, 114}, extent = {{10, -10}, {-10, 10}}, rotation = 0)));
+  // initial parameters
```
done

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/PowerExternalBase.mo:
```diff
@@ -0,0 +1,27 @@
+*/
+
+block PowerExternalBase "Outputs u1^u2"
+
```
done

martinscripts commented on nrt/data/SMIB/Standard/TestCaseGovCT2/TestCaseGovCT2.jobs:
I added it, please check if it's correct

joyelfeghali commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,357 @@
+  parameter Types.PerUnit ValveMaxPu = 1 "Maximum valve position limit in pu (base PNomTurb*KTurb), typical: 1" annotation( Dialog(tab = "Turbine/engine"));
+  parameter Types.PerUnit ValveMinPu = 0.175 "Minimum valve position limit in pu (base PNomTurb*KTurb), typical: 0.175" annotation( Dialog(tab = "Turbine/engine"));
+  parameter Types.PerUnit WFnlPu = 0.187 "No load fuel flow in pu (base PNomTurb*KTurb), typical: 0.187" annotation( Dialog(tab = "Turbine/engine"));
+  parameter Boolean WFSpdBool = false "Switch for fuel source characteristic, typical: false" annotation( Dialog(tab = "Turbine/engine")); 
```
Yes it's disturbing in dynawo, we have seen problems when imposing a default value in the Modelica model, the change in the .par won't be taken into account.
That's why the parameters' value should be given in the .par only. 
since we put a typical value in the definition of each parameter, this can be useful to the user that is using it in OpenModelica.

martinscripts replied:
(I did add them now. please verify.)

martinscripts replied:
okay, I will remove the default values

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,272 @@
+  // initial parameters
+  parameter Types.AngularVelocityPu omega0Pu = SystemBase.omega0Pu "Initial rotor speed";
+  final parameter Types.ActivePowerPu PRef0Pu = RDroop*(if RSelectInt == 0 then 0 elseif RSelectInt == 1 then Pm0Pu else initValvePu) "Initial value of reference power";
+  parameter Types.ActivePowerPu PGen0Pu = Pm0Pu*PBaseMw/PGenBaseMw "Initial value of generated active power in pu (base SnRef (system base), generator convention";
```
the preassembled uses 
`initName="Dynawo.Electrical.Controls.Machines.Governors.GovernorPmPGen_INIT"`

This defines PGen0Pu as a signal. 

How do I fix this? Shall I just remove PGen0Pu from the INIT, so that the definition in GovCT2.mo is used instead?

joyelfeghali replied:
No just remove the definition in GovCT2.mo because the one coming from the generator should be used

martinscripts replied:
ah, now I got you

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,272 @@
+    // initialization helpers
+    final parameter Types.PerUnit initCfePu = WFnlPu + (if KTurb > 0 then initPMechNoLossPu/KTurb else 0) "Initial value of cfe in pu";
+    final parameter Types.PerUnit initFsrtPu = (PLdRefPu/KTurb + WFnlPu - initTexPu)*KPLoad + initIntegratorKILoadPu "Initial value of cfe in fsrt in pu";
+    final parameter Types.PerUnit initIntegratorKILoadPu = 1 "Initial value of load controller integrator in pu";
```
It is initialized to a large value, so that it is not the active path at init.

since the limiter's upper limit is 1 (see screenshot), the largest meaningful value is 1.

![image](https://github.com/user-attachments/assets/651b648f-69d9-43d1-9012-32a48ad5aaaa)

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/PowerExternalBase.mo:
```diff
@@ -0,0 +1,27 @@
+* This file is part of Dynawo, an hybrid C++/Modelica open source suite of simulation tools for power systems.
+*/
+
+block PowerExternalBase "Outputs u1^u2"
```
You were right, this works with Power.mo 
I removed the PowerExternalBase model

martinscripts replied:
Yes, that's what I meant. I adapted the model now.

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,272 @@
+  Modelica.Blocks.Interfaces.RealInput omegaRefPu(start = SystemBase.omegaRef0Pu) "Rotor speed set-point in pu (base omegaNom)" annotation(Placement(visible = true, transformation(origin = {-333, 23}, extent = {{-13, -13}, {13, 13}}, rotation = 0), iconTransformation(origin = {-343, 181}, extent = {{-23, -23}, {23, 23}}, rotation = 0)));
+  Modelica.Blocks.Interfaces.RealInput PGenPu(start = PGen0Pu) "Generated electrical active power in pu (base SnRef (system base)) (generator convention)" annotation( Placement(visible = true, transformation(origin = {-335, -183}, extent = {{-13, -13}, {13, 13}}, rotation = 0), iconTransformation(origin = {-343, -161}, extent = {{-23, -23}, {23, 23}}, rotation = 0)));
+  Modelica.Blocks.Interfaces.RealOutput PmPu(start = Pm0Pu) "Mechanical power output in pu (base PNomTurb) (generator convention)" annotation( Placement(visible = true, transformation(origin = {330, 132}, extent = {{-10, -10}, {10, 10}}, rotation = 0), iconTransformation(origin = {350, 2}, extent = {{-30, -30}, {30, 30}}, rotation = 0)));
+  Modelica.Blocks.Interfaces.RealInput PMwSetPu(start = PRef0Pu) "Supervisory power controller set-point / automatic generation control (base PNomTurb) (generator convention)" annotation( Placement(visible = true, transformation(origin = {-333, -139}, extent = {{-13, -13}, {13, 13}}, rotation = 0), iconTransformation(origin = {-344, -80}, extent = {{-24, -24}, {24, 24}}, rotation = 0)));
```
I think you're correct

martinscripts replied:
Ah, in my testcase I am setting it to Pm0Pu, which also works. 
This was a bug, here.

joyelfeghali replied:
But PGen0Pu * PGenBaseMw / PBaseMw  is not equivalent to Pm0Pu

martinscripts replied:
yes, but I was not sure which of them would be better in this case. 
I will change it.

martinscripts replied:
done

martinscripts commented:
I added changes according to your review. Please have a look. I think I didn't miss any conversations this time, but correct me if I'm wrong.

martinscripts commented:
I added 
`<par type="DOUBLE" name="governor_PGenBaseMw" value="100"/>`

to  .par, 
can I instead use the parameter SystemBase.SnRef somehow or just leave 100?

martinscripts commented:
apart from the open conversations, I added the requested changes

joyelfeghali commented:
> I added `<par type="DOUBLE" name="governor_PGenBaseMw" value="100"/>`
> 
> to .par, can I instead use the parameter SystemBase.SnRef somehow or just leave 100?

you can leave it 100

joyelfeghali commented:
> I added `<par type="DOUBLE" name="governor_PGenBaseMw" value="100"/>`
> 
> to .par, can I instead use the parameter SystemBase.SnRef somehow or just leave 100?

Yes you can keep it 100.

martinscripts commented:
I added a SMIB test case. 

It does not run. 
I think the backward initialization of omega into the generator fails. 

error:
![image](https://github.com/user-attachments/assets/6c387567-917a-4e06-b347-2a9aebcdd89d)

If I don't connect omegaPu and just give a constant to the governor's omega input, then it works. 

and transformational debugger points to omega: 
![image](https://github.com/user-attachments/assets/475af23d-957a-4af4-bf11-cdf8355bb834)


I just pushed the smib test case. 

do you have any ideas?

joyelfeghali commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,352 @@
+  Dynawo.NonElectrical.Blocks.Continuous.PowerExternalBase expOmegaToTheDm annotation( Placement(visible = true, transformation(origin = {214, 150}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
+  Dynawo.NonElectrical.Blocks.Continuous.RateLimFirstOrderFreeze firstOrdertActuatorRatelim(T = tActuatorSeconds, UseRateLim = true, Y0 = initValve, y(fixed = true)) annotation( Placement(visible = true, transformation(origin = {214, -88}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
+  Modelica.Blocks.Continuous.FirstOrder firstOrdertFLoad(T = tFLoadSeconds, initType = Modelica.Blocks.Types.Init.InitialOutput, y_start = initTex) annotation( Placement(visible = true, transformation(origin = {34, 114}, extent = {{10, -10}, {-10, 10}}, rotation = 0)));
+  Modelica.Blocks.Continuous.FirstOrder firstOrdertPElec(T = tPElecSeconds, initType = Modelica.Blocks.Types.Init.InitialOutput, y_start = Pm0Pu) annotation( Placement(visible = true, transformation(origin = {-250, -182}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
```
Here y_start = PGen0Pu not to Pm0Pu

martinscripts replied:
done

joyelfeghali commented on dynawo/sources/Models/Modelica/PreassembledModels/CMakeLists.txt:
```diff
@@ -63,6 +63,9 @@ set(MODELICA_PREASSEMBLED_MODELS
   GeneratorSynchronousFourWindingsGovSteam1St4bPss2b.xml
+  GeneratorSynchronousFourWindingsGovCT2SexsPss2A.xml
+  GeneratorSynchronousFourWindingsGovSteam1ExcIEEEST4B.xml
+  GeneratorSynchronousFourWindingsGovSteam1ExcIEEEST4BPssIEEE2B.xml
```
Also remove they were renamed
  GeneratorSynchronousFourWindingsGovSteam1ExcIEEEST4B.xml
  GeneratorSynchronousFourWindingsGovSteam1ExcIEEEST4BPssIEEE2B.xml

joyelfeghali commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/package.mo:
```diff
@@ -1,20 +1,22 @@
+  */
 
 package Standard "Standard governors"
+
```
no line skip

joyelfeghali commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/package.mo:
```diff
@@ -1,20 +1,22 @@
+  * SPDX-License-Identifier: MPL-2.0
+  *
+  * This file is part of Dynawo, an hybrid C++/Modelica open source suite of simulation tools for power systems.
+  */
```
no indentation from line 3 to line 13

joyelfeghali commented on dynawo/sources/Models/Modelica/Dynawo/Examples/SMIB/Standard/package.order:
```diff
@@ -1,3 +1,7 @@
 GovSteam1St4bPss2b3
+GovSteam1ExcIEEEST4BPssIEEE2B1
+GovSteam1ExcIEEEST4BPssIEEE2B2
+GovSteam1ExcIEEEST4BPssIEEE2B3
```
also here remove 
GovSteam1ExcIEEEST4BPssIEEE2B1
GovSteam1ExcIEEEST4BPssIEEE2B2
GovSteam1ExcIEEEST4BPssIEEE2B3

joyelfeghali commented on dynawo/sources/Models/Modelica/Dynawo/Examples/SMIB/Standard/package.order:
```diff
@@ -1,3 +1,7 @@
+GovSteam1ExcIEEEST4BPssIEEE2B1
+GovSteam1ExcIEEEST4BPssIEEE2B2
+GovSteam1ExcIEEEST4BPssIEEE2B3
+GovCT2ExcIEEEST4BPssIEEE2B2
```
rename this model here too

joyelfeghali commented on nrt/data/SMIB/Standard/TestCaseGovCT2/reference/outputs/curves.csv:
the folder where curves.csv is in should be : reference/outputs/curves/curves.csv

martinscripts replied:
we had this before, didn't we? I am sure that I changed this and I wonder why the change disappeared

joyelfeghali replied:
Yes, now there is just one folder missing

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,480 @@
+    Placement(visible = true, transformation(origin = {-335, -183}, extent = {{-13, -13}, {13, 13}}, rotation = 0), iconTransformation(origin = {-343, -161}, extent = {{-23, -23}, {23, 23}}, rotation = 0)));
+  Modelica.Blocks.Interfaces.RealOutput PmPu(start = Pm0Pu) "Mechanical power output in pu (base PNomTurb) (generator convention)" annotation(
+    Placement(visible = true, transformation(origin = {330, 132}, extent = {{-10, -10}, {10, 10}}, rotation = 0), iconTransformation(origin = {350, 2}, extent = {{-30, -30}, {30, 30}}, rotation = 0)));
+  Modelica.Blocks.Interfaces.RealInput PMwSetPu(start = PGen0Pu * PGenBaseMw / PBaseMw) "Supervisory power controller set-point / automatic generation control (base PNomTurb) (generator convention)" annotation(
```
done

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/Governors/Standard/Generic/GovCT2.mo:
```diff
@@ -0,0 +1,480 @@
+    Placement(visible = true, transformation(origin = {80, 114}, extent = {{10, -10}, {-10, 10}}, rotation = 0)));
+  
+  // initial parameters
+  parameter Types.AngularVelocityPu omega0Pu "Initial rotor speed in pu (base omegaNom), typical value = SystemBase.omega0Pu";
```
done, I removed the parameter omega0Pu

martinscripts commented on dynawo/sources/Models/Modelica/Dynawo/Examples/SMIB/Standard/GovCT2ExcIEEEST4BPssIEEE2B2.mo:
```diff
@@ -0,0 +1,211 @@
+    __OpenModelica_simulationFlags(initialStepSize = "0.001", lv = "LOG_STATS", nls = "kinsol", s = "ida", nlsLS = "klu", maxIntegrationOrder = "2", maxStepSize = "10", emit_protected = "()"),
+    Diagram(coordinateSystem(extent = {{-160, -100}, {160, 100}})),
+  Documentation(info = "<html><head></head><body><span style=\"font-family: 'DejaVu Sans Mono'; font-size: 12px;\">This test case has been created to test the GovCT2 governor model.&nbsp;</span><div><span style=\"font-family: 'DejaVu Sans Mono'; font-size: 12px;\">It has been derifved from&nbsp;</span>Dynawo.Examples.SMIB.Standard.GovSteam1ExcIEEEST4BPssIEEE2B2.</div><div><br></div><div><span sty ...
+    <figure>
```
done
